src/lib/dnssd/IncrementalResolve.cpp

```diff
@@ -341,17 +344,36 @@ CHIP_ERROR IncrementalResolver::OnIpAddress(Inet::InterfaceId interface, const I
     return CHIP_NO_ERROR;
 }
 
+template <class NodeType>
+CHIP_ERROR IncrementalResolver::Set(DiscoveredNodeData & output)
+{
+    VerifyOrReturnError(mSpecificResolutionData.Is<NodeType>(), CHIP_ERROR_INCORRECT_STATE);
+    output.Set<NodeType>();
+    auto & nodeData                       = output.Get<NodeType>();
+    nodeData                              = mSpecificResolutionData.Get<NodeType>();
+    CommonResolutionData & resolutionData = nodeData;
+    resolutionData                        = mCommonResolutionData;
+    return CHIP_NO_ERROR;
```

Comment:
There are several intermediate variables above. Could you re-visit the logic and see if it can be cleaned up? 

As it stands, this feels a bit too much magic in casting and operator=. Why do we have this complex of code?

Reply:
At a minimum `output.Set<NodeType>(mSpecificResolutionData.Get<NodeType>())` makes some of the assignments more explicity ... however I seem to get lost in what mSpecificResolutionData vs output are supposed to contain and their difference in types.

Reply:
I see this was copied & pasted from previous code, however previous code was also part of the refactor. It used to be:

```cpp
outputData.resolutionData = mCommonResolutionData;
outputData.nodeData       = mSpecificResolutionData.Get<DnssdNodeData>();
```

however now it seems significantly messier. This needs fixing to be readable.

Reply:
changed this a bit.